File: clients/node-sdk/src/BaseClient.ts
/**
 * Base client with common HTTP functionality
 */

import axios, { AxiosInstance, AxiosRequestConfig, AxiosResponse } from 'axios';
import {
  RoutingDispatchError,
  AuthenticationError,
  ValidationError,
  NotFoundError,
  RateLimitError,
} from './errors';
import { SDKConfig, ErrorResponse } from './types';

export class BaseClient {
  protected axiosInstance: AxiosInstance;

  constructor(config: SDKConfig) {
    const baseURL = config.baseURL || 'http://localhost:3000/api';
    const timeout = config.timeout || 30000;

    this.axiosInstance = axios.create({
      baseURL,
      timeout,
      headers: {
        'Authorization': `Bearer ${config.apiKey}`,
        'Content-Type': 'application/json',
        'User-Agent': 'routing-dispatch-node-sdk/1.0.0',
      },
    });

    // Response interceptor for error handling
    this.axiosInstance.interceptors.response.use(
      (response) => response,
      (error) => {
        if (error.response) {
          const { status, data } = error.response;
          const errorData = data as ErrorResponse;

          switch (status) {
            case 401:
              throw new AuthenticationError('Invalid API key');

            case 400:
              throw new ValidationError(
                errorData.message || 'Validation failed',
                errorData.errors
              );

            case 404:
              throw new NotFoundError(errorData.message || 'Not found');

            case 429:
              const retryAfter = error.response.headers['retry-after'];
              throw new RateLimitError(
                'Rate limit exceeded',
                retryAfter ? parseInt(retryAfter, 10) : undefined
              );

            default:
              throw new RoutingDispatchError(
                errorData.message || 'API error',
                status
              );
          }
        } else if (error.request) {
          throw new RoutingDispatchError('Network error: No response received');
        } else {
          throw new RoutingDispatchError(`Request error: ${error.message}`);
        }
      }
    );
  }

  protected async get<T>(endpoint: string, params?: any): Promise<T> {
    const response: AxiosResponse<T> = await this.axiosInstance.get(endpoint, {
      params,
    });
    return response.data;
  }

  protected async post<T>(endpoint: string, data?: any): Promise<T> {
    const response: AxiosResponse<T> = await this.axiosInstance.post(
      endpoint,
      data
    );
    return response.data;
  }

  protected async put<T>(endpoint: string, data?: any): Promise<T> {
    const response: AxiosResponse<T> = await this.axiosInstance.put(
      endpoint,
      data
    );
    return response.data;
  }

  protected async delete<T>(endpoint: string): Promise<T> {
    const response: AxiosResponse<T> = await this.axiosInstance.delete(endpoint);
    return response.data;
  }
}

File: clients/python-sdk/routing_dispatch_sdk/base_client.py
"""Base client with common HTTP functionality"""

import requests
from typing import Dict, Any, Optional
from .exceptions import (
    RoutingDispatchError,
    AuthenticationError,
    ValidationError,
    NotFoundError,
    RateLimitError,
)


class BaseClient:
    """Base client for API communication"""

    def __init__(
        self,
        api_key: str,
        base_url: str = "http://localhost:3000/api",
        timeout: int = 30,
    ):
        self.api_key = api_key
        self.base_url = base_url.rstrip("/")
        self.timeout = timeout
        self.session = requests.Session()
        self.session.headers.update(
            {
                "Authorization": f"Bearer {api_key}",
                "Content-Type": "application/json",
                "User-Agent": "routing-dispatch-python-sdk/1.0.0",
            }
        )

    def _request(
        self,
        method: str,
        endpoint: str,
        data: Optional[Dict[str, Any]] = None,
        params: Optional[Dict[str, Any]] = None,
    ) -> Dict[str, Any]:
        """
        Make HTTP request to API

        Args:
            method: HTTP method (GET, POST, PUT, DELETE)
            endpoint: API endpoint path
            data: Request body data
            params: Query parameters

        Returns:
            Response data as dictionary

        Raises:
            AuthenticationError: When authentication fails
            ValidationError: When request validation fails
            NotFoundError: When resource not found
            RateLimitError: When rate limit exceeded
            RoutingDispatchError: For other API errors
        """
        url = f"{self.base_url}/{endpoint.lstrip('/')}"

        try:
            response = self.session.request(
                method=method,
                url=url,
                json=data,
                params=params,
                timeout=self.timeout,
            )

            # Handle error responses
            if response.status_code == 401:
                raise AuthenticationError("Invalid API key")
            elif response.status_code == 400:
                error_data = response.json() if response.content else {}
                raise ValidationError(
                    error_data.get("message", "Validation failed"),
                    errors=error_data.get("errors"),
                )
            elif response.status_code == 404:
                raise NotFoundError(response.json().get("message", "Not found"))
            elif response.status_code == 429:
                retry_after = response.headers.get("Retry-After")
                raise RateLimitError(
                    "Rate limit exceeded",
                    retry_after=int(retry_after) if retry_after else None,
                )
            elif response.status_code >= 400:
                error_msg = response.json().get("message", "API error")
                raise RoutingDispatchError(error_msg, status_code=response.status_code)

            response.raise_for_status()

            # Return JSON response
            return response.json() if response.content else {}

        except requests.exceptions.Timeout:
            raise RoutingDispatchError(f"Request timeout after {self.timeout}s")
        except requests.exceptions.ConnectionError:
            raise RoutingDispatchError(f"Connection error to {url}")
        except requests.exceptions.RequestException as e:
            raise RoutingDispatchError(f"Request failed: {str(e)}")

    def get(self, endpoint: str, params: Optional[Dict] = None) -> Dict:
        """Make GET request"""
        return self._request("GET", endpoint, params=params)

    def post(self, endpoint: str, data: Optional[Dict] = None) -> Dict:
        """Make POST request"""
        return self._request("POST", endpoint, data=data)

    def put(self, endpoint: str, data: Optional[Dict] = None) -> Dict:
        """Make PUT request"""
        return self._request("PUT", endpoint, data=data)

    def delete(self, endpoint: str) -> Dict:
        """Make DELETE request"""
        return self._request("DELETE", endpoint)
